ibis/expr/timecontext.py

```diff
@@ -127,6 +127,88 @@ def canonicalize_context(
     return begin, end
 
 
+def construct_time_context_aware_series(
+    series: pd.Series, frame: pd.DataFrame
+) -> pd.Series:
+    """ Construct a Series by adding 'time' in its MultiIndex
+
+    In window execution, the result Series of udf may need
+    to be trimmed by timecontext. In order to do so, 'time'
+    must be added as an index to the Series. We extract
+    time column from the parent Dataframe `frame`.
+    See `trim_with_timecontext` in execution/window.py for
+    trimming implementation.
+
+    Examples:
+        Assume frame is:
+                             time  id  value
+        0 2017-01-02 01:02:03.234   1    1.1
+        1 2017-01-03 01:02:03.234   2    2.2
+        2 2017-01-04 01:02:03.234   3    3.3
+
+        For a series of:
+        0    1.1
+        1    2.2
+        2    3.3
+        Name: value, dtype: float64
+
+        The result series will be:
+                             time
+        0 2017-01-02 01:02:03.234     1.1
+        1 2017-01-03 01:02:03.234     2.2
+        2 2017-01-04 01:02:03.234     3.3
+        Name: value, dtype: float64
+        The index will be a MultiIndex of the original RangeIndex
+        and a DateTimeIndex.
+
+        For a series already has 'time' as its index:
+                             time
+        2017-01-02 01:02:03.234     1.1
+        2017-01-03 01:02:03.234     2.2
+        2017-01-04 01:02:03.234     3.3
+        Name: value, dtype: float64
+        The result is unchanged.
+
+        For a series with MultiIndex but 'time' is not present in
+        the MultiIndex:
+          id
+        0  1     1.1
+        1  2     2.2
+        2  3     3.3
+        Name: value, dtype: float64
+        'time' will be added into the MultiIndex in result:
+          id                     time
+        0  1  2017-01-02 01:02:03.234     1.1
+        1  2  2017-01-03 01:02:03.234     2.2
+        2  3  2017-01-04 01:02:03.234     3.3
+        Name: value, dtype: float64
+
+    Parameters
+    ----------
+    series: pd.Series, the result series of an udf execution
+    frame: pd.DataFrame, the parent Dataframe of `series`
+
+    Returns
+    -------
+    pd.Series
+    """
+    if TIME_COL == frame.index.name:
+        time_index = frame.index
+    elif TIME_COL in frame:
+        time_index = frame.set_index(TIME_COL).index
+    else:
+        raise com.IbisError(f'"time" column not present in DataFrame {frame}')
+    if TIME_COL not in series.index.names:
+        series.index = pd.MultiIndex.from_arrays(
```

Comment:
You probably want Jeff to take a look at this code. I don't know if this is the best way.

Reply:
Does the ordering of indexes matter? e.g., key, time vs time, key

Reply:
Yeah, this is similar to implementation in [window_agg_built_in](https://github.com/ibis-project/ibis/blob/08261f53f7685f44754f52741e025473aecef762/ibis/backends/pandas/aggcontext.py#L430)

Reply:
In terms of trimming, no, when we trim by timecontext MultiIndex is flatten, in [trim_with_timecontext](https://github.com/ibis-project/ibis/blob/08261f53f7685f44754f52741e025473aecef762/ibis/backends/pandas/execution/window.py#L195)

Reply:
Gotcha